Merge/pathfinding buildingsystem

CodyBol opened the pull request.

CodyBol committed:
Created the basis for a grid system

CodyBol committed:
Added a build atlas

CodyBol committed:
Added a tilemap update system

CodyBol committed:
Added a multi select

CodyBol committed:
Added a hover

CodyBol committed:
Added multiple shape size

CodyBol committed:
Added cell groups

CodyBol committed:
Added rotation and multiple burshes

CodyBol committed:
Added a changeable brush

CodyBol committed:
Created all the data scrtuctures

CodyBol committed:
Updated to SubBuildItems

CodyBol committed:
Added walls

CodyBol committed:
added some documentation

CodyBol committed:
updated the cursor

CodyBol committed:
Added better rotation

CodyBol committed:
Added new assets

joost2000 committed:
Pathfinding almost finished (see description))

- Pathfinding works on cpu without multi threading but is too slow.
- Working on unity jobs implementation of it.
-  Code is a messs because this is my first time working with jobs, will clean it up when its up and running

CodyBol committed:
Fixed flatten array

CodyBol committed:
Added simple UI

CodyBol committed:
added scene

CodyBol committed:
Added a nicer dropdown

CodyBol committed:
Updated the drag brush

joost2000 committed:
Finished first version of pathfinding

CodyBol committed:
Updated cursor to demo

CodyBol committed:
Updated to demos

joost2000 committed:
Created a seperate NPC controller

CodyBol committed:
Added new scene

CodyBol committed:
rider files

CodyBol committed:
Merge branch 'feature/building-system' into feature/pathfinding

# Conflicts:
#	.idea/.idea.Airport-Tycoon/.idea/.gitignore
#	Assets/Scenes/SampleScene.unity
#	Packages/manifest.json

CodyBol committed:
Building system merge

CodyBol committed:
added scene

joost2000 committed:
Fixed bug that caused pathfinding to break

NPC still doesn't fully reach it's path, might have something to do with grid offset, look into this

joost2000 committed:
Made NPC fully reach end of path

Issue: Path is not always optimal, issue somewhere in Astar struct

CodyBol committed:
Added zone

CodyBol committed:
Added grass background

joost2000 committed:
Cleaned up files

CodyBol committed:
Merge branch 'feature/building-system' into merge/pathfinding-buildingsystem

# Conflicts:
#	Assets/Scenes/BuildingSystem.unity

CodyBol committed:
Changed the order of files

CodyBol committed:
updated meta files

CodyBol committed:
Merge branch 'development' into merge/pathfinding-buildingsystem

# Conflicts:
#	Assets/Scenes/CameraDemoScene.unity

CodyBol committed:
Moved audio files to demo

CodyBol committed:
moved files

CodyBol committed:
Added buildable tiles

CodyBol committed:
Merge branch 'development' into merge/pathfinding-buildingsystem

# Conflicts:
#	Assets/Scenes/AudioSampleScene.unity.meta

CodyBol committed:
Added BuilderWorker and GameManager

CodyBol committed:
Added Camera controls

CodyBol committed:
Added the event manager

CodyBol committed:
Updated EventManager

joost2000 committed:
Merged worker and new movement

joost2000 committed:
Merge branch 'merge/pathfinding-buildingsystem' of https://github.com/HvA-Aviation/Airport-Tycoon into merge/pathfinding-buildingsystem

CodyBol committed:
Updated namespace

TJJonker commented on Assets/Features/Building/Scripts/Datatypes/BuildableObject.cs:
```diff
@@ -0,0 +1,14 @@
+    public class BuildableObject : ScriptableObject
+    {
+        public string Name;
+        //public Tile Sprite;
```
Remove commented code if no longer necessary

TJJonker commented on Assets/Features/Building/Scripts/Datatypes/BuildableObject.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace Features.Building.Scripts.Datatypes
+{
+    [CreateAssetMenu(fileName = "Building/BuildItem", menuName = "Building/BuildItem", order = 0)]
```
Filenames rarely contain slashes

TJJonker commented on Assets/Features/Building/Scripts/Datatypes/BuildableObject.cs:
```diff
@@ -0,0 +1,14 @@
+namespace Features.Building.Scripts.Datatypes
+{
+    [CreateAssetMenu(fileName = "Building/BuildItem", menuName = "Building/BuildItem", order = 0)]
+    public class BuildableObject : ScriptableObject
```
If everything is public, might as well turn it into a struct

CodyBol replied:
ScriptableObject are not allowed to be structs

TJJonker commented on Assets/Features/Building/Scripts/Datatypes/SubBuildItem.cs:
```diff
@@ -0,0 +1,26 @@
+            GridPosition = new GridPosition(new Vector2Int(gridPosition.x, gridPosition.y), (GridLayer)gridPosition.z);
+        }
+
+        public Tile Tile;
```
Be consistent with where you put your variables and functions. Variables at the top of your script and functions underneath them is the common approach you use in the rest of your scripts

TJJonker replied:
Public functions form a threat of being changed directly. Since your seemingly using functions to change them, but want them publicly accessible to read from, try using properties: `Public Tile Tile {get; private set;}`

TJJonker commented on Assets/Features/Building/Scripts/Datatypes/TileData.cs:
```diff
@@ -0,0 +1,13 @@
+namespace Features.Building.Scripts.Datatypes
+{
+    [Serializable]
+    public class TileData
```
All public, making it a struct might look nicer

CodyBol replied:
struct is not allowed in inspector so I added properties

TJJonker commented on Assets/Features/Building/Scripts/Datatypes/TileData.cs:
```diff
@@ -0,0 +1,13 @@
+    {
+        public Tile Tile;
+        public bool UnTraversable = false;
+        //TODO different tile for rotation list
```
Remove comment and place it in the backlog if necessary

TJJonker commented on Assets/Features/Building/Scripts/Grid/CellData.cs:
```diff
@@ -0,0 +1,29 @@
+namespace Features.Building.Scripts.Grid
+{
+    [Serializable]
+    public class CellData
```
Struct might be more suitable since everything is public, or think about protecting variables where necessary

TJJonker commented on Assets/Features/Building/Scripts/Grid/CellData.cs:
```diff
@@ -0,0 +1,29 @@
+        //TODO change to workload
+        public float BuildPercentage = .4f;
+
+        public bool Build(float speed)
```
Still not sure about this one, as we discussed... But alternative either, so leave it for now

TJJonker replied:
About the implemented use of time*

CodyBol replied:
I added this to the backlog and this will be done in the future

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+        [SerializeField] private Color _validColor;
+        [SerializeField] private Color _invalidColor;
+
+        [SerializeField] private List<SubBuildItem> _selectedGroup = new List<SubBuildItem>();
```
Seems to be a private variable without the need to be visible in the editor. Might as well remove the SerializeField tag

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+        [SerializeField] private Color _invalidColor;
+
+        [SerializeField] private List<SubBuildItem> _selectedGroup = new List<SubBuildItem>();
+        [SerializeField] private List<SubBuildItem> _shape = new List<SubBuildItem>();
```
Same for this one, might as well remove the SerializeField tag

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+
+        [SerializeField] private List<SubBuildItem> _selectedGroup = new List<SubBuildItem>();
+        [SerializeField] private List<SubBuildItem> _shape = new List<SubBuildItem>();
+        [SerializeField] private Vector3Int _origin;
```
And this one

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+        [SerializeField] private List<SubBuildItem> _selectedGroup = new List<SubBuildItem>();
+        [SerializeField] private List<SubBuildItem> _shape = new List<SubBuildItem>();
+        [SerializeField] private Vector3Int _origin;
+        [SerializeField] private int _rotation = 0;
```
And this one

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+        [SerializeField] private int _rotation = 0;
+
+        [SerializeField] private EventSystem _eventSystem;
+        [SerializeField] private BuildableObject _selectedBuilding;
```
And this one

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+
+        private void Start()
+        {
+            _grid = FindObjectOfType<Grid>();
```
Might as well use a SerializeField for this reference

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+
+            transform.localScale = Vector3.one * _grid.CellSize;
+
+            ChangeToBuilding(_selectedBuilding);
```
I don't think you should save the default selected (building) in the cursor, but in a buildingmanager instead. Ultimately, the cursor shouldn't even care what it is holding or visualizing. This building manager could also manage which buildings are locked and unlocked, which will be necessary for the R&D tree and progression

TJJonker replied:
Oh and be more clear with your variable naming. `_selectedBuilding` in this context should be named `_defaultSelectedBuilding` or something along these lines. Preferrably without the 'Building' part, cause you're not always building a building. Could also be furniture, which practically speaking is no building ;P

CodyBol replied:
Added a ticket on atlassian and moved this to a demo script

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+
+        private void Update()
+        {
+            var pos = Camera.main.ScreenToWorldPoint(Input.mousePosition);
```
Try to avoid using `var`. Be clear with your variables for readability

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+
+            _tilemap.gameObject.SetActive(!_eventSystem.IsPointerOverGameObject());
+
+            if (!_tilemap.gameObject.activeSelf)
```
Checking whether the tilemap is active? Shouldn't it always be active? Making this check redundant...?

CodyBol replied:
There is a difference between the normal tilemap and the cursor tilemap. The cursor tile map is something on is own and is rendered on a whole other layer. This makes sure that the draw functionality of the cursor is disabled when over an UI element

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+            if (!_tilemap.gameObject.activeSelf)
+                return;
+
+            if (Input.GetKey(KeyCode.B))
```
Hardcoded input. Move this outside the script and into a demo script

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+                return;
+            }
+
+            if (Input.GetMouseButton(1) || Input.GetMouseButtonUp(1))
```
Same for this, hardcoded input, move this to a demo script and work with public functions

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+            }
+            else
+            {
+                switch (_selectedBuilding.BrushType)
```
Could be moved into separate brush scripts as discussed

CodyBol replied:
Has been placed in the backlog

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+            {
+                Vector2Int dir = Vector2Int.one;
+
+                if (direction == 1)
```
Do 180 degree rotations also work or are we limited to 90 degs? Seems like you only counted on inputs being able to be -1, 0 and 1, right?

CodyBol replied:
not for now, but atm we don't need them, but I'll add a ticket

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+                    _shape[i].GridPosition.Position = new Vector2Int(gridPosition.y, gridPosition.x) * dir;
+                }
+
+                if (_rotation > 3)
```
Could change these if statements into a single `_rotation %= 4;` or `_rotation = _rotation % 4;`

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+
+        private void MultiBrushSelect(Vector3Int position, int cursor)
+        {
+            if (Input.GetMouseButtonDown(cursor))
```
Hardcoded input, move this to a demo script and use public functions

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+                //starts selection and sets the origin of the selection
+                _origin = position;
+            }
+            else if (Input.GetMouseButtonUp(cursor))
```
Same here

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+            else if (Input.GetMouseButtonUp(cursor))
+            {
+                //place or remove. This ends the selection
+                if (cursor == 0)
```
Unclear what `cursor` is. Try adjusting the name to be more clear

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+                _selectedGroup = new List<SubBuildItem>();
+                _tilemap.ClearAllTiles();
+            }
+            else if (Input.GetMouseButton(cursor) && _selectedGroup.Count > 0)
```
Again, hardcoded input

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+            {
+                //gets selection
+                //get the min and the max of the position
+                var min = Vector3Int.Min(_origin, position);
```
Don't use `var`. Maybe using `Rect` is also an option for this?

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+                List<SubBuildItem> currentSelectedGroup = new List<SubBuildItem>();
+
+                //get all tiles between the min and the max position
+                for (var x = min.x; x < max.x + 1; x++)
```
`var` = no bueno

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+
+        #region brushes
+
+        private void MultiBrushSelect(Vector3Int position, int cursor)
```
Try to get rid of this huge if tree. Makes it hard to read and prone for errors

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+
+        private void OutlineBrushSelect(Vector3Int position)
+        {
+            if (Input.GetMouseButtonDown(0))
```
Hardcoded input

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+            {
+                _origin = position;
+            }
+            else if (Input.GetMouseButtonUp(0))
```
Hardcoded input

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+                    }
+
+                //resets the selection 
+                _selectedGroup = new List<SubBuildItem>();
```
Could also use the `Clear()` function instead of creating a new list

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+                _selectedGroup = new List<SubBuildItem>();
+                _tilemap.ClearAllTiles();
+            }
+            else if (Input.GetMouseButton(0) && _selectedGroup.Count > 0)
```
Hardcoded input

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+
+                _grid.SetGroup(positions, indices, _rotation);
+            }
+            else if (Input.GetMouseButtonDown(1))
```
Uh oh, hardcode input, no bueno

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+
+        private void SingleBrush(Vector3Int position)
+        {
+            if (Input.GetMouseButtonDown(0))
```
Another one

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+        {
+            Hover(position, _shape);
+
+            if (Input.GetMouseButton(0))
```
Another one

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+            {
+                _grid.Set(position, _selectedBuilding.BuildItems[0].Tile);
+            }
+            else if (Input.GetMouseButton(1))
```
And another one

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+
+        private void Hover(Vector3Int position, List<SubBuildItem> shapes, bool requireAllAvailable = false, bool flipColors = false)
+        {
+            //Sets the offset of the whole grid
```
Be consistent with the use of capitalized letters at the beginning of the sentence in your comments

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+                }
+            }
+
+            var validColor = !flipColors ? _validColor : _invalidColor;
```
Replace the `var`

TJJonker replied:
For the next two occasions as well

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+                TileChangeData tempTile = new TileChangeData()
+                {
+                    position = gridPosition.GridPosition - offset,
+                    color = valid
```
Difficult to read with the double ternary operators

CodyBol replied:
Any recommendations?

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+            _selectedBuilding = buildableObject;
+
+            //Create a fresh shape so the rotation is correct and doesn't change the prefab
+            _shape = new List<SubBuildItem>();
```
Could also be cleared instead of making a new list

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+            _selectedGroup = currentSelectedGroup;
+        }
+
+        public void ChangeToBuilding(BuildableObject buildableObject)
```
`ChangeToBuilding()` might not be the right name for this function since we're not changing to a building, instead changing the object to build. Something like `ChangeSelectedBuildable()` might be more suitable.

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,367 @@
+            }
+        }
+
+        public void DisableCursor()
```
Not sure why disabling the cursor hides the tilemap instead. Same for enabling the cursor

CodyBol replied:
Seperate Cursor tilemap

TJJonker commented on Assets/Features/Building/Scripts/Grid/Grid.cs:
```diff
@@ -0,0 +1,391 @@
+        }
+
+        /// <summary>
+        /// Creates a flattend 2d array to see if it is traversable
```
'To see if it is traversable'. What is it checking?

TJJonker commented on Assets/Features/Building/Scripts/Grid/Grid.cs:
```diff
@@ -0,0 +1,391 @@
+        /// Creates a flattend 2d array to see if it is traversable
+        /// </summary>
+        /// <returns>A flattend 2d bool array with false as traversable</returns>
+        public bool[,] UnTraversable()
```
A name Like `GetUntraversable()` might be more suitable, although it might be more straightforward to return a 2d array containing traversable tiles instead of non traversable tiles. At least, when reading it

TJJonker commented on Assets/Features/Building/Scripts/Grid/Grid.cs:
```diff
@@ -0,0 +1,391 @@
+
+        /// <summary>
+        /// Loops through the whole grid and sets all the cells to what they are supposed to be
+        /// TODO add a buffer, so correct tiles aren't changed
```
Put this in he backlog if necessary, but remove the comment since it will get lost in the big codebase.
I know you can look for TODO tags, but keeping this in the backlog is easier

TJJonker commented on Assets/Features/Building/Scripts/Grid/Grid.cs:
```diff
@@ -0,0 +1,391 @@
+                {
+                    for (int z = 0; z < _gridSize.z; z++)
+                    {
+                        var cell = _cells[x, y, z];
```
`var` is a nono

TJJonker commented on Assets/Features/Building/Scripts/Grid/Grid.cs:
```diff
@@ -0,0 +1,391 @@
+            if (Get(gridVector) == -1)
+                return true;
+        
+            var buildTiles = new List<Vector3Int>() { gridVector };
```
`var`

TJJonker commented on Assets/Features/Building/Scripts/Grid/Grid.cs:
```diff
@@ -0,0 +1,391 @@
+            return isFinished;
+        }
+
+        /// <summary>
```
Be consistent with documentation, either do everything or nothing. Rather have the first one though

TJJonker commented on Assets/Features/Building/Scripts/Grid/Grid.cs:
```diff
@@ -0,0 +1,391 @@
+        {
+            if (Get(gridVector) == -1)
+            {
+                var cellData = _cells[gridVector.x, gridVector.y, gridVector.z];
```
`var`

TJJonker commented on Assets/Features/Building/Scripts/Grid/Grid.cs:
```diff
@@ -0,0 +1,391 @@
+            if (!OutOfBounds(gridVector) && Get(gridVector) != -1)
+            {
+                //checks if given tile is part of a linked group. If not only add the given position
+                var group = _cellGroup.FirstOrDefault(x => x.Any(j => j == gridVector));
```
`var`

TJJonker commented on Assets/Features/Building/Scripts/Grid/Grid.cs:
```diff
@@ -0,0 +1,391 @@
+                {
+                    _cells[item.x, item.y, item.z].Tile = -1;
+                    _cells[item.x, item.y, item.z].Rotation = 0;
+                    _cells[item.x, item.y, item.z].BuildPercentage = 0.4f;
```
`Buildpercentage` is hardcoded?

CodyBol replied:
For the time being yes. Just so we have a quick visualisation. The build percentage is directly coupled to the opacity

TJJonker commented on Assets/Features/Building/Scripts/Grid/Grid.cs:
```diff
@@ -0,0 +1,391 @@
+            {
+                for (int y = 0; y < _gridSize.y; y++)
+                {
+                    //var origin = new Vector2(x, y) * _cellSize - (new Vector2(_gridSize.x, _gridSize.y) * _cellSize / 2 - Vector2.one * _cellSize / 2);
```
Remove commented code

TJJonker commented on Assets/Features/Building/Scripts/Grid/GridPosition.cs:
```diff
@@ -0,0 +1,39 @@
+            Layer = layer;
+        }
+
+        public Vector2Int Position;
```
Seems like you don't want other classes to be able to change these two values. Make them properties instead. Plus if all content is public, consider making it a struct

CodyBol replied:
Changed this to a private Tile with public Tile Getter. With public tile {get; private set} it won't be shown in the editor

TJJonker commented on Assets/Features/EventManager/EventId.cs:
Emtpy class?

CodyBol replied:
Made this a enum used for the event system

TJJonker commented on Assets/Features/Managers/GameManager.cs:
```diff
@@ -0,0 +1,18 @@
+    {
+        public static GameManager Instance;
+        
+        public AudioManager.Scripts.AudioManager AudioManager;
```
Weird reference to the AudioManager `AudioManager.Scripts.AudioManager`. Might be nicer to just say `AudioManager` instead

CodyBol replied:
Added a better reference in using. This is because of the already existing autio manager of unity

TJJonker commented on Assets/Features/Managers/GameManager.cs:
```diff
@@ -0,0 +1,18 @@
+        public static GameManager Instance;
+        
+        public AudioManager.Scripts.AudioManager AudioManager;
+        public EventManager EventManager;
```
Properties might be better since we can now create a new instance and reassign it

CodyBol replied:
Added properties again with them seperated, becuase of the inspector issue

TJJonker commented on Assets/Implementation/Workers/BuilderTaskSystem.cs:
Seems like overkill to create a full class for just a single variable which could also be put inline in the gamemanager

CodyBol replied:
Moved this directly to the TaskManager. Maybe later create a dictionary or something when we have multiple taskSystems

CodyBol committed:
Finished some of the feedback

CodyBol committed:
Changed SubItem back to class, but with protected variables

TJJonker commented on Assets/Implementation/Pathfinding/Scripts/NPCController.cs:
```diff
@@ -0,0 +1,213 @@
+    [SerializeField] private bool _drawGizmos;
+
+    [Header("Movement variables")]
+    [SerializeField] float _moveSpeed = 1f;
```
Don't forget to define the protection. Public, private, protected etc

TJJonker commented on Assets/Implementation/Pathfinding/Scripts/NPCController.cs:
```diff
@@ -0,0 +1,213 @@
+
+    public Vector3Int GetGridPosition()
+    {
+        var pos = transform.position;
```
Don't use `var`

TJJonker commented on Assets/Implementation/Pathfinding/Scripts/NPCController.cs:
```diff
@@ -0,0 +1,213 @@
+        var pos = transform.position;
+
+        //clamp to grid positions
+        var clampedValue = new Vector2(RoundToMultiple(pos.x, _grid.CellSize),
```
Don't use `var`

TJJonker commented on Assets/Implementation/Pathfinding/Scripts/NPCController.cs:
```diff
@@ -0,0 +1,213 @@
+        var clampedValue = new Vector2(RoundToMultiple(pos.x, _grid.CellSize),
+            RoundToMultiple(pos.y, _grid.CellSize));
+
+        if (Input.GetKeyDown(KeyCode.P) & _pathCompleted)
```
Hardcoded input control. Move this to a demo file and control it with public functions

TJJonker commented on Assets/Implementation/Pathfinding/Scripts/NPCController.cs:
```diff
@@ -0,0 +1,213 @@
+            }
+            transform.position = path[i].position;
+        }
+        _pathCompleted = true;
```
Think it's cooler and better to use a callback that executes when the path is complete. This way, you don't have to constantly poll whether the path is done. Instead, the function is executed when the path is completed

TJJonker commented on Assets/Implementation/Pathfinding/Scripts/NPCController.cs:
```diff
@@ -0,0 +1,213 @@
+                Node node = new Node()
+                {
+                    position = new Vector3Int(x, y, 0),
+                    traversable = !untraversable[x, y]
```
This is something from the Grid system, but reading the line `traversable = !untraversable[x, y]` would be much easier to read if it just was  `IsTraversable = traversable[x, y]` or something along these lines.

TJJonker commented on Assets/Implementation/Pathfinding/Scripts/NPCController.cs:
```diff
@@ -0,0 +1,213 @@
+
+    private void OnDrawGizmosSelected()
+    {
+        if (!_drawGizmos) return;
```
I don't see the point of having a bool that determines whether or not gizmos are drawn when the only moment they're drawn is when the object is selected...

CodyBol committed:
Updated changes

CodyBol committed:
Updated a lot of the reviewed items

CodyBol committed:
progress

CodyBol committed:
Updated so it uses work speed system

CodyBol committed:
Added the brushes of a useless morning

CodyBol committed:
Created the start of the new brushes

CodyBol committed:
Added outline brush

CodyBol committed:
Changed to the new brush system

joost2000 committed:
Updated some PR stuff check description

- Removed not used code
- defined variable protections
- removed serializefield things where they were not needed
- chanjged traversable to untraversable to better fit cody's variable names
-  still working on callback for the pathcompleted but will be done today

CodyBol committed:
Added a remove tool

CodyBol committed:
Changed remove tool settings to buildable

CodyBol committed:
Merge branch 'refs/heads/rework/brushes' into merge/pathfinding-buildingsystem

CodyBol committed:
Finished merging

CodyBol committed:
Merge branch 'refs/heads/feature/work-amount' into merge/pathfinding-buildingsystem

CodyBol committed:
Changed to traversable

CodyBol committed:
CHanged to %=

CodyBol committed:
removed the double ternary operator

CodyBol committed:
Placed traversable tiles in a seperate variable

joost2000 committed:
Made callback for when the destionation is reached

CodyBol committed:
added uncommited change of workload

TJJonker commented on Assets/Features/Building/Scripts/Brushes/MultiBrush.cs:
```diff
@@ -0,0 +1,51 @@
+        {
+            //gets selection
+            //get the min and the max of the position
+            var min = Vector3Int.Min(_origin, position);
```
Don't use `var` here

TJJonker commented on Assets/Features/Building/Scripts/Brushes/MultiBrush.cs:
```diff
@@ -0,0 +1,51 @@
+            _selectedTiles.Clear();
+
+            //get all tiles between the min and the max position
+            for (var x = min.x; x < max.x + 1; x++)
```
Be consistent with the use of `var` in loops. Here you use `var`, In OutlineBrush.cs you use int in for loops.

TJJonker commented on Assets/Features/Building/Scripts/Grid/Cursor.cs:
```diff
@@ -0,0 +1,198 @@
+            bool valid = true;
+            if (requireAllAvailable)
+            {
+                foreach (var gridPosition in selectedGroup)
```
Rather use Vec2 / Vec3 here instead of `var` since I don't know which of the two this is.

CodyBol committed:
Updated the vars
